docs/proposals/08-shoot-additional-container-runtimes.md

```diff
@@ -0,0 +1,159 @@
+    1. Add kubelet execution flags:
+        1. --container-runtime=remote
+        2. --container-runtime-endpoint=unix:///.../containerd.socket
+    2. Make sure that containerd configuration file exist in path /etc/containerd/config.toml and it is configured to enable the CRI plugin (In the containerd config.toml "cri" plugin must not be disabled).
+    
+2. Containerd socket path and containerd configurations are different for each OS. To make sure the configurations above are set in each worker machine, each OS extension would be responsible to configure them.
+    1. os-ubuntu - 
+        1. Add a controlplane webhook to manipulate the OperatingSystemConfig and add the following flags to the kubelet execution command:
+            1. --container-runtime=remote
+            2. --container-runtime-endpoint=unix:///run/containerd/containerd.sock
+        2. Create a containerd configuration file: /etc/containerd/config.toml based on the current configuration file in the path: __TBD__ and remove the CRI disabled plugin.
+    2. os-coreos - 
+        1. Add a controlplane webhook to manipulate the OperatingSystemConfig and add the following flags to the kubelet execution command:
+           1. --container-runtime=remote
+           2. --container-runtime-endpoint=unix:///var/run/docker/libcontainerd/docker-containerd.socket
+        2. Create a containerd configuration file: /etc/containerd/config.toml based on the default configuration of containerd. 
+        3. Add a generator that will override the default containerd.service unit. It will create a new unit in: /etc/systemd/system/containerd.service unit that will start the containerd service with a configuration file path: /etc/containerd/config.toml.
+        
+    3. os-suse-jeos - __TBD__ 
+    
+3. Docker-monitor daemon and rotate log should be replaced with equivalent conatinerd services. __TBD__ 
+4. In order to install additional runtimes on the worker machine we must provide the installation scripts to the machine file system and run them.
+   The installation script should download and install the runtime binaries in each OS.
+   Instructions for Gardener to copy files to the machine filesystem and execute them is done via the cloud config secret.
+   
+   1. Each OS extension should implement a generator that will copy a generic packages installation script file.
+        ```yaml
+          #cloud-config
+          write_files:
+            path: '/var/lib/packages/install.sh'
+            ...
+      ```
+      
+      The installation file will be executed with different parameters defining the relevant package that should be installed:
+        1. Package name
+        2. Package path url
+        3.  __TBD__ 
+        
+   2. Each Extension controller should reconcile the OperatingSystemConfig (Should be enhanced to contain additional runtime data __TBD__):
```

Comment:
You mean each CRI extension controller? They should not reconcile the `OperatingSystemConfig` resource as the OSC controllers are responsible for doing this. Instead, they could inject additional service units via a webhook?

Reply:
I will change it to be implemented via a CRI webhook. Do you think that we should extend the `OperatingSystemConfig` to contain the available runtime or use another resource?
We do need this information in every worker machine bootstrapping (As well as in a cluster level).

Reply:
> Do you think that we should extend the OperatingSystemConfig to contain the available runtime or use another resource?

Hm, not sure yet, I have to come back. The problem is that OSC and CRI is very bundled.